src/backend/distributed/utils/metadata_cache.c

```diff
@@ -194,13 +195,68 @@ IsDistributedTable(Oid relationId)
 		return false;
 	}
 
-	cacheEntry = LookupDistTableCacheEntry(relationId);
+	/*
+	 * We will not use cache if there is a version mismatch, because while
+	 * creating the cache, we assume our metadata tables are in particular
+	 * structure. However, in version mismatch, this is not guaranteed, thus
+	 * we will read from pg_dist_partition table and we will only assume
+	 * specific details which are most likely do not change in long term.
+	 */
+	installedVersion = InstalledExtensionVersion();
+	if (!MajorVersionsCompatible(installedVersion, CITUS_EXTENSIONVERSION))
+	{
+		if (IsDistributedTableFromCatalog(relationId))
+		{
+			ereport(ERROR, (errmsg("it is not allowed to perform operations on "
+								   "distributed tables when library and extension "
+								   "versions mismatch"),
+							errdetail("Loaded library requires %s, but the installed "
+									  "extension version is %s.", CITUS_MAJORVERSION,
+									  installedVersion),
+							errhint("Run ALTER EXTENSION citus UPDATE and try again.")));
+		}
+
+		return false;
+	}
 
+	cacheEntry = LookupDistTableCacheEntry(relationId);
 	return cacheEntry->isDistributedTable;
 }
 
 
 /*
+ * IsDistributedTableFromCatalog searches pg_dist_partition table for given
+ * relationId and returns true if it founds given relationId in the table.
+ * It specifically does not use distributed cache, because this function is
+ * designed to be used in cases where accessing metadata tables is not safe.
+ */
+static bool
+IsDistributedTableFromCatalog(Oid relationId)
+{
+	HeapTuple partitionTuple = NULL;
+	SysScanDesc scanDescriptor = NULL;
+	const int scanKeyCount = 1;
+	ScanKeyData scanKey[scanKeyCount];
+	bool indexOK = true;
+
+	Relation pgDistPartition = heap_open(DistPartitionRelationId(), AccessShareLock);
+
+	ScanKeyInit(&scanKey[0], Anum_pg_dist_partition_logicalrelid,
+				BTEqualStrategyNumber, F_OIDEQ, ObjectIdGetDatum(relationId));
+
+	scanDescriptor = systable_beginscan(pgDistPartition,
+										DistPartitionLogicalRelidIndexId(),
+										indexOK, NULL, scanKeyCount, scanKey);
+
+	partitionTuple = systable_getnext(scanDescriptor);
```

Comment:
I'm a bit doubtful that it's a good idea to rely on DistPartitionRelationId(), DistPartitionLogicalRelidIndexId() etc here. But we can possibly defer that concern to a later point.